auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+                }
+
+                if (send_tok.length != 0) {
+                    if (const auto err = krb_send_token( &send_tok, fd ); !err.ok()) {
+                        // TODO: should we be checking these release calls?
+                        gss_release_buffer(&minorStatus, &send_tok);
+                        gss_release_name(&minorStatus, &target_name);
+                        THROW(err.code(), err.result());
+                    }
+                }
+
+                gss_release_buffer(&minorStatus, &send_tok);
+
+                if (majorStatus == GSS_S_CONTINUE_NEEDED) {
+                    recv_tok.value = &ikrbScratchBuffer;
+                    recv_tok.length = SCRATCH_BUFFER_SIZE;
+                    unsigned int bytes_read;
+                    if (const auto ret = krb_receive_token(fd, &recv_tok, &bytes_read); !ret.ok()) {
+                        gss_release_name( &minorStatus, &target_name );
+                        const auto err = PASSMSG("Error reading KRB token.", ret);
+                        THROW(err.code(), err.result());
+                    }
+
+                    tokenPtr = &recv_tok;
+                }
+            }
+            while (majorStatus == GSS_S_CONTINUE_NEEDED);
+
+            resp["next_operation"] = AUTH_CLIENT_AUTH_RESPONSE;
+
+            return resp;
+        } // kerberos_auth_establish_context
+
+        json kerberos_auth_client_response(rcComm_t& comm, const json& req)
+        {
+            namespace ie = irods::experimental;
+
+            if (auto [e, n] = verify(req, {"user_name","zone_name"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in request", __func__, __LINE__, n));
+            }
+
+            json svr_req{req};
+            svr_req["next_operation"] = AUTH_AGENT_AUTH_RESPONSE;
+            auto resp = request(comm, svr_req);
+
+            comm.loggedIn = 1;
+
+            resp["next_operation"] = ie::auth::flow_complete;
+
+            return resp;
+        } // kerberos_auth_client_response
+
+#ifdef RODS_SERVER
+        json kerberos_auth_agent_request(rsComm_t& comm, const json& req)
+        {
+            if (krbAuthReqStatus == 1) {
+                krbAuthReqStatus = 0;
+                if (krbAuthReqError != 0) {
+                    rodsLogAndErrorMsg(LOG_NOTICE, &comm.rError, krbAuthReqError, krbAuthReqErrorMsg);
+                    THROW(krbAuthReqError, "A KRB auth request error has occurred.");
+                }
+            }
+
+            const auto kerberos_name = irods::get_server_property<std::string>("KerberosServicePrincipal");
+
+            constexpr bool client = false;
+            const auto service_name = krb_setup_creds(&comm.rError, client, kerberos_name);
+
+            comm.gsiRequest = 1;
+            if (comm.auth_scheme) {
+                free(comm.auth_scheme);
+            }
+            comm.auth_scheme = strdup(irods::AUTH_KRB_SCHEME.c_str());
```

Comment:
Does this result in a leak?

Can be investigated later.

Reply:
I will make an issue. This logic exists in all of our auth plugin implementations.

Reply:
Created https://github.com/irods/irods/issues/6375 for this